traitlets/config/application.py

```diff
@@ -581,18 +581,31 @@ def _load_config_files(cls, basefilename, path=None, log=None, raise_config_file
         yield each config object in turn.
         """
 
+        def new_loader(name, path):
+            if name.endswith('.py'):
+                return cls.python_config_loader_class(name, path=path, log=log)
+            elif name.endswith('.json'):
+                return cls.json_config_loader_class(name, path=path, log=log)
+
         if not isinstance(path, list):
             path = [path]
         for path in path[::-1]:
             # path list is in descending priority order, so load files backwards:
-            pyloader = cls.python_config_loader_class(basefilename+'.py', path=path, log=log)
             if log:
                 log.debug("Looking for %s in %s", basefilename, path or os.getcwd())
-            jsonloader = cls.json_config_loader_class(basefilename+'.json', path=path, log=log)
+            pyloader = new_loader(basefilename + '.py', path=path)
+            jsonloader = new_loader(basefilename + '.json', path=path)
+            loaders = [pyloader, jsonloader]
```

Comment:
I have a question not strictly related to the core functionality of this PR
Why did you changed the separate `cls.XXX_config_loader_class()` calls with `new_loader()`?

I also need to change the abov 3 lines,
but for adding a new loader (i.e. YAML) without overriding and reimplementing `_load_config_files()`,
like that:
```diff
-            pyloader = cls.python_config_loader_class(basefilename+'.py', path=path, log=log)
             if log:
                 log.debug("Looking for %s in %s", basefilename, path or os.getcwd())
-            jsonloader = cls.json_config_loader_class(basefilename+'.json', path=path, log=log)
+           loaders = cls._make_loaders(basefilename, path=path, log=log)
+           log.debug("Looking for %s in %s", basefilename, path or os.getcwd())
```

You may implement `_make_loaders()` either with `new_loader()` or `cls.XXX_config_loader_class()` calls.
 Then adding a new loader requires only to overrd this trivial method.